src/lib/Parser.hs

```diff
@@ -336,20 +336,23 @@ decl = do
 interfaceInstance :: Parser UDecl
 interfaceInstance = do
   keyWord InstanceKW
-  (lhs, pos) <- withPos simpleLet
+  (p, pos) <- withPos $ letPat <|> parens pat
+  ann <- annot uType
   keyWord WhereKW
-  rhs <- withSrc $ URecord <$> interfaceRecordFields "="
-  let ULet _ (_, Just (WithSrc _  ann)) _ = lhs rhs
-  let ULet _ (p, ann') rhs' = lhs $ mkRhs ann rhs
-  let (ann'', rhs'') = addImplicitImplicitArgs pos ann' rhs'
-  return $ ULet InstanceLet (p, ann'') rhs''
+  rhs <- mkRhs ann <$> (withSrc $ URecord <$> interfaceRecordFields "=")
+  let (ann', rhs') = addImplicitImplicitArgs pos (Just ann) rhs
+  return $ ULet InstanceLet (p, ann') rhs'
   where
-    mkRhs (UPi _ _ (WithSrc _ lrhs)) rhs = mkRhs lrhs rhs
-    mkRhs (UApp _ (WithSrc _ (UVar v)) _) rhs =
-      (var . nameToStr . mkInterfaceConsName . varName) v `mkApp` rhs
-    -- We let this case through to fail during type checking.
-    mkRhs _ rhs = rhs
-    var s = WithSrc Nothing $ UVar $ mkName s :> ()
+    mkRhs (WithSrc _ ann) = mkRhs' ann
```

Comment:
You keep unpacking `WithSrc` in every single case of `mkRhs'`, so maybe just flatten it into a single `mkRhs` function that always starts by unpacking the `WithSrc`?